File: app/src/main/java/com/wolfsea/pulltorefreshdemo/loadrefreshlayout/RecyclerRefreshLayout.java
package com.wolfsea.pulltorefreshdemo.loadrefreshlayout;
import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.wolfsea.pulltorefreshdemo.loadrefreshlayout.base.RefreshLayout;

public class RecyclerRefreshLayout extends RefreshLayout {

    private RecyclerView mRecyclerView;

    public RecyclerRefreshLayout(Context context) {
        this(context, null);
    }

    public RecyclerRefreshLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (mRecyclerView == null) {

            findListView(this);
        }
    }

    private void findListView(View view) {
        if (view instanceof ViewGroup) {
            ViewGroup viewGroup = (ViewGroup) view;
            if (viewGroup.getChildCount() > 0) {
                for (int i = 0; i < viewGroup.getChildCount(); i++) {

                    View viewItem = viewGroup.getChildAt(i);

                    if (viewItem instanceof RecyclerView) {

                        mRecyclerView = (RecyclerView) viewItem;
                        if (mRecyclerView.getAdapter() != null
                                && mRecyclerView.getAdapter() instanceof LoadRecyclerAdapter) {

                            ((LoadRecyclerAdapter) mRecyclerView.getAdapter()).setFootView(footView);
                        }

                        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
                            @Override
                            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                                super.onScrolled(recyclerView, dx, dy);
                                if (isAutoLoad) {

                                    loadData();
                                }
                            }

                            @Override
                            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                                super.onScrollStateChanged(recyclerView, newState);
                            }
                        });
                        return;
                    } else {

                        findListView(viewItem);
                    }
                }
            }
        }
    }

    /**
     * 判断是否到了最底部
     */
    @Override
    protected boolean isReachBottom() {
        if (mRecyclerView == null) {

            return false;
        } else {

            LinearLayoutManager lm = (LinearLayoutManager) mRecyclerView.getLayoutManager();
            int position = lm.findLastVisibleItemPosition();
            int count = lm.getItemCount();
            return position > count - 2;
        }
    }

    @Override
    protected void showLoadView(boolean isShow) {
        if (mRecyclerView != null && mRecyclerView.getAdapter() != null
                && mRecyclerView.getAdapter() instanceof LoadRecyclerAdapter) {

            LoadRecyclerAdapter adapter = (LoadRecyclerAdapter) mRecyclerView.getAdapter();
            if (isShow) {

                adapter.showFootView(true);
            } else {

                adapter.showFootView(false);
            }
        }
    }

}

File: app/src/main/java/com/wolfsea/pulltorefreshdemo/loadrefreshlayout/RecyclerRefreshLayout.kt
package com.wolfsea.pulltorefreshdemo.loadrefreshlayout
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.wolfsea.pulltorefreshdemo.loadrefreshlayout.base.RefreshLayout

/**
 *@desc  RecyclerRefreshLayout
 *@author liuliheng
 *@time 2021/7/7  23:50
 **/
class RecyclerRefreshLayout : RefreshLayout {

    private var recyclerView: RecyclerView? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attributeSet: AttributeSet) : super(context, attributeSet)

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        if (recyclerView == null) {
            findView(this)
        }
    }

    override fun isReachBottom(): Boolean {
        return if (recyclerView == null) {

            false
        } else {

            val layoutManager = recyclerView?.layoutManager as LinearLayoutManager
            val position = layoutManager.findLastVisibleItemPosition()
            val count = layoutManager.itemCount
            position > count - 2
        }
    }

    override fun showLoadView(isShow: Boolean) {
        recyclerView?.adapter?.let {
            if (it is LoadRecyclerAdapter) {
                it.showFootView(isShow)
            }
        }
    }

    private fun findView(view: View) {
        if (view is ViewGroup) {
            for (i in 0.until(view.childCount)) {

                val child = getChildAt(i)
                if (child is RecyclerView) {

                    recyclerView = child
                    if (recyclerView?.adapter!! is LoadRecyclerAdapter) {

                        (child.adapter as LoadRecyclerAdapter).setFootView(footView)
                    }

                    recyclerView?.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                        override fun onScrollStateChanged(
                            recyclerView: RecyclerView,
                            newState: Int
                        ) {
                            super.onScrollStateChanged(recyclerView, newState)
                        }

                        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                            super.onScrolled(recyclerView, dx, dy)
                            if (isAutoLoad) {

                                loadData()
                            }
                        }
                    })
                    break
                } else {

                    findView(child)
                }
            }
        }
    }

}